### PR TITLE
Fixes for `java.util.HashSet` and `java.util.LinkedHashSet`

### PR DESCRIPTION
**Pull request by alexeev509**
1. action SYMBOLIC was replaced by MAP_GET_ANY_KEY
2. Was added method toString();

### CODE REVIEW

File: spec/java/util/HashSet.KeyIterator.lsl [hunk 1]
```diff
@@ -89,12 +89,9 @@ automaton HashSet_KeyIteratorAutomaton
...
             action THROW_NEW("java.util.NoSuchElementException", []);
 
-        val key: Object = action SYMBOLIC("java.lang.Object");
-        action ASSUME(key != null);
-        action ASSUME(key != this.currentKey);
         val parentStorage: map<Object, Object> = HashSetAutomaton(this.parent).storage;
-        val sourceStorageHasKey: boolean = action MAP_HAS_KEY(parentStorage, key);
-        action ASSUME(sourceStorageHasKey);
+        val key: Object = action MAP_GET_ANY_KEY(parentStorage);
+        action ASSUME(key != this.currentKey);
```
Comment (dog-m) [inline spec/java/util/HashSet.KeyIterator.lsl:94]: Consider using supplementary `unseen: map<Object, Object>` to store items for later retrieval as suggested by Valentin.

File: spec/java/util/HashSet.main.lsl [hunk 2]
```diff
@@ -127,10 +128,7 @@ automaton HashSetAutomaton
...
     proc _generateKey (visitedKeys: map<Object, Object>): Object
     {
-        result = action SYMBOLIC("java.lang.Object");
-        action ASSUME(result != null);
-        val isKeyExist: boolean = action MAP_HAS_KEY(this.storage, result);
-        action ASSUME(isKeyExist);
+        result = action MAP_GET_ANY_KEY(this.storage);
 
         val isKeyWasVisited: boolean = action MAP_HAS_KEY(visitedKeys, result);
         action ASSUME(!isKeyWasVisited);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: Consider using supplementary `unseen: map<Object, Object>` to store items for later retrieval as suggested by Valentin.

File: spec/java/util/LinkedHashSet.KeyIterator.lsl [hunk 3]
```diff
@@ -88,12 +88,9 @@ automaton LinkedHashSet_KeyIteratorAutomaton
...
 
-        val key: Object = action SYMBOLIC("java.lang.Object");
-        action ASSUME(key != null);
-        action ASSUME(key != this.currentKey);
         val parentStorage: map<Object, Object> = LinkedHashSetAutomaton(this.parent).storage;
-        val sourceStorageHasKey: boolean = action MAP_HAS_KEY(parentStorage, key);
-        action ASSUME(sourceStorageHasKey);
+        val key: Object = action MAP_GET_ANY_KEY(parentStorage);
+        action ASSUME(key != this.currentKey);
         val dstStorageHasKey: boolean = action MAP_HAS_KEY(this.visitedKeys, key);
```
Comment (dog-m) [inline spec/java/util/LinkedHashSet.KeyIterator.lsl]: Consider using supplementary `unseen: map<Object, Object>` to store items for later retrieval as suggested by Valentin.

File: spec/java/util/LinkedHashSet.main.lsl [hunk 4]
```diff
@@ -124,10 +125,7 @@ automaton LinkedHashSetAutomaton
 
     proc _generateKey (visitedKeys: map<Object, Object>): Object
     {
-        result = action SYMBOLIC("java.lang.Object");
-        action ASSUME(result != null);
-        val isKeyExist: boolean = action MAP_HAS_KEY(this.storage, result);
-        action ASSUME(isKeyExist);
+        result = action MAP_GET_ANY_KEY(this.storage);
 
         val isKeyWasVisited: boolean = action MAP_HAS_KEY(visitedKeys, result);
```
Comment (dog-m) [inline spec/java/util/LinkedHashSet.main.lsl]: Consider using supplementary `unseen: map<Object, Object>` to store items for later retrieval as suggested by Valentin.

**Review by dog-m — CHANGES_REQUESTED** [review]

File: spec/java/util/HashSet.KeyIterator.lsl (hunk 1 above)
Comment (alexeev509) [reply spec/java/util/HashSet.KeyIterator.lsl:94]: Fixed

File: spec/java/util/HashSet.main.lsl (hunk 2 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed

File: spec/java/util/LinkedHashSet.KeyIterator.lsl (hunk 3 above)
Comment (alexeev509) [reply spec/java/util/LinkedHashSet.KeyIterator.lsl]: Fixed

File: spec/java/util/LinkedHashSet.main.lsl (hunk 4 above)
Comment (alexeev509) [reply spec/java/util/LinkedHashSet.main.lsl]: Fixed

File: spec/java/util/HashSet.main.lsl [hunk 5]
```diff
@@ -250,10 +246,11 @@ automaton HashSetAutomaton
 
     fun *.iterator (@target self: HashSet): Iterator
     {
-        val visitedKeysMap: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 6]
```diff
@@ -125,15 +126,10 @@ automaton HashSetAutomaton
     }
 
 
-    proc _generateKey (visitedKeys: map<Object, Object>): Object
+    @KeepVisible proc _generateKey (unseenKeys: map<Object, Object>): Object
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: If this subroutine has became so simple and straightforward, can we completely discard it while moving its contents into appropriate call sites?

File: spec/java/util/HashSet.main.lsl [hunk 7]
```diff
@@ -285,11 +282,13 @@ automaton HashSetAutomaton
     fun *.spliterator (@target self: HashSet): Spliterator
     {
         val keysStorageArray: array<Object> = action ARRAY_NEW("java.lang.Object", this.length);
-        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 8]
```diff
@@ -372,10 +370,12 @@ automaton HashSetAutomaton
         }
         else
         {
-            val visitedKeys: map<Object, Object> = action MAP_NEW();
+            val unseenKeys: map<Object, Object> = action MAP_NEW();
+            action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 9]
```diff
@@ -421,25 +420,24 @@ automaton HashSetAutomaton
         val len: int = this.length;
         result = action ARRAY_NEW("java.lang.Object", len);
         val expectedModCount: int = this.modCount;
-        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 10]
```diff
@@ -448,7 +446,8 @@ automaton HashSetAutomaton
         val expectedModCount: int = this.modCount;
         val aLen: int = action ARRAY_SIZE(a);
         val len: int = this.length;
-        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 11]
```diff
@@ -476,12 +475,13 @@ automaton HashSetAutomaton
         val len: int = this.length;
         result = action CALL(generator, [0]) as array<Object>;
         val expectedModCount: int = this.modCount;
-        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 12]
```diff
@@ -568,11 +568,12 @@ automaton HashSetAutomaton
         val lengthBeforeAdd: int = this.length;
         val expectedModCount: int = this.modCount;
         var i: int = 0;
-        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/HashSet.main.lsl [hunk 13]
```diff
@@ -612,25 +612,25 @@ automaton HashSetAutomaton
 
         var i: int = 0;
         val expectedModCount: int = this.modCount;
-        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/HashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

File: spec/java/util/LinkedHashSet.main.lsl [hunk 14]
```diff
@@ -122,15 +123,10 @@ automaton LinkedHashSetAutomaton
     }
 
 
-    proc _generateKey (visitedKeys: map<Object, Object>): Object
+    @KeepVisible proc _generateKey (unseenKeys: map<Object, Object>): Object
```
Comment (dog-m) [inline spec/java/util/LinkedHashSet.main.lsl]: Can we discard and manually inline this subroutine?

File: spec/java/util/LinkedHashSet.main.lsl [hunk 15]
```diff
@@ -241,10 +237,11 @@ automaton LinkedHashSetAutomaton
 
     fun *.iterator (@target self: LinkedHashSet): Iterator
     {
-        val visitedKeysMap: map<Object, Object> = action MAP_NEW();
+        val unseenKeys: map<Object, Object> = action MAP_NEW();
+        action MAP_UNITE_WITH(unseenKeys, this.storage);
```
Comment (dog-m) [inline spec/java/util/LinkedHashSet.main.lsl]: The `MAP_CLONE` action may be helpful here.

**Review by dog-m — CHANGES_REQUESTED** [review]

File: spec/java/util/HashSet.main.lsl (hunk 5 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed

File: spec/java/util/HashSet.main.lsl (hunk 7 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

File: spec/java/util/HashSet.main.lsl (hunk 8 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

File: spec/java/util/HashSet.main.lsl (hunk 9 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

File: spec/java/util/HashSet.main.lsl (hunk 10 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

File: spec/java/util/HashSet.main.lsl (hunk 11 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

File: spec/java/util/HashSet.main.lsl (hunk 13 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

File: spec/java/util/LinkedHashSet.main.lsl (hunk 15 above)
Comment (alexeev509) [reply spec/java/util/LinkedHashSet.main.lsl]: Fixed;

File: spec/java/util/LinkedHashSet.main.lsl (hunk 14 above)
Comment (alexeev509) [reply spec/java/util/LinkedHashSet.main.lsl]: Fixed;

File: spec/java/util/HashSet.main.lsl (hunk 6 above)
Comment (alexeev509) [reply spec/java/util/HashSet.main.lsl]: Fixed;

**Review by dog-m — APPROVED** [review]

**Pull request merged**